## Project Detail
- Data wrangling, which consists of:
    - Gathering data
    - Assessing data
    - Cleaning data
- Storing, analyzing, and visualizing our wrangled data
- Reporting on 1) our data wrangling efforts and 2) our data analyses and visualizations

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import requests
import json
import time
import os
import re

## Gather Data

### Twitter Archive

In [2]:
# Read a csv file as a Pandas DataFrame
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [3]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

### Image Predictions

In [4]:
# Download image_predictions.tsv file from Udacity server
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(url)

In [5]:
with open ('image_predictions.tsv' ,'wb') as file:
    file.write(response.content)
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')    

In [6]:
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [7]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


 ### Tweets

In [8]:
# Keep only certain elements in dataframe
elements_to_keep = ['id', 'favorite_count', 'retweet_count']
data = []

with open('tweet_json.txt', 'r') as file: 
    # Read the line from JSON.txt and convert it into dataframe
    tweet_json = file.readline()
    
    while tweet_json:
        tweet_dict = json.loads(tweet_json)
        # Create a smaller dict
        data_row = dict((element, tweet_dict[element]) for element in elements_to_keep)
        data.append(data_row)
        
        # Read the next line
        tweet_json = file.readline()
            
df_tweet= pd.DataFrame.from_dict(data)

In [9]:
df_tweet.head()

favorite_count                  id  retweet_count
0           39467  892420643555336193           8853
1           33819  892177421306343426           6514
2           25461  891815181378084864           4328
3           42908  891689557279858688           8964
4           41048  891327558926688256           9774

In [10]:
df_tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
favorite_count    2354 non-null int64
id                2354 non-null int64
retweet_count     2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


## Assess Data

In [11]:
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [12]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [13]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [14]:
twitter_archive.name.value_counts()

None        745
a            55
Charlie      12
Cooper       11
Oliver       11
Lucy         11
Tucker       10
Penny        10
Lola         10
Bo            9
Winston       9
the           8
Sadie         8
Toby          7
an            7
Buddy         7
Daisy         7
Bailey        7
Rusty         6
Dave          6
Koda          6
Scout         6
Jax           6
Leo           6
Jack          6
Stanley       6
Oscar         6
Bella         6
Milo          6
Sammy         5
           ... 
Snickers      1
Anthony       1
Maya          1
Travis        1
Erik          1
Ginger        1
Danny         1
Ike           1
Storkson      1
Eevee         1
Monster       1
Coleman       1
Jordy         1
Chesney       1
Willie        1
Ashleigh      1
Sweet         1
Ralphy        1
Mollie        1
Gin           1
Kayla         1
Comet         1
Amy           1
Miley         1
Finnegus      1
Lassie        1
Howie         1
General       1
Bode          1
Carter        1
Name: name, Length: 957,

In [15]:
twitter_archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [16]:
twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [17]:
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [18]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [19]:
df_tweet.head()

favorite_count                  id  retweet_count
0           39467  892420643555336193           8853
1           33819  892177421306343426           6514
2           25461  891815181378084864           4328
3           42908  891689557279858688           8964
4           41048  891327558926688256           9774

In [20]:
df_tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
favorite_count    2354 non-null int64
id                2354 non-null int64
retweet_count     2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


## Quality Issues
### twitter_archive
- Wrong data type: **timestamp** and **retweeted_status_timestamp** should be datetime instead of object.
- Wrong data type: **in_reply_to_status_id** ,**in_reply_to_user_id**, **retweeted_status_id**, and **retweeted_status_user_id** should be string instead of float64.
- Retweets: Some entries are retweets. We only want to keep the original.
- Name: In **name** column, many entries do not look like names.
- Rating: The **numerator** and **denominator** columns have strange values.
- Remove columns which are not required for the analysis and store the dataframe to twitter_archive_master.csv

### image_predictions
- Inconsistent capitalization: Some of the first letter in p1/p2/p3 are capital.
- Missing values: 2075 rows instead of 2356

### df_tweet
- Rename **id** column to **tweet_id** to be consistent with **twitter_archive** and **image_predictions**

## Tidiness Issues
- Join **image_predictions** and **df_tweet** to **twitter_archive**.
- **twitter_archive**: one variable in four columns (doggo, floofer, pupper, puppo).

## Clean Data

In [21]:
# Copy dataframes
df_clean_twitter_archive = twitter_archive.copy()
df_clean_image_predictions = image_predictions.copy()
df_clean_tweet = df_tweet.copy()

#### Rename **id** column to **tweet_id** to be consistent with **twitter_archive** and **image_predictions**

In [22]:
df_clean_tweet.rename(columns={'id' : 'tweet_id'}, inplace=True)
df_clean_tweet.columns

Index(['favorite_count', 'tweet_id', 'retweet_count'], dtype='object')

#### Join **image_predictions** and **df_tweet** to **twitter_archive**.

In [23]:
df_clean_twitter_archive = pd.merge(df_clean_twitter_archive, df_clean_image_predictions, on='tweet_id', how='inner')
df_clean_twitter_archive = pd.merge(df_clean_twitter_archive, df_clean_tweet, on='tweet_id', how='inner')
df_clean_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 0 to 2072
Data columns (total 30 columns):
tweet_id                      2073 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2073 non-null object
source                        2073 non-null object
text                          2073 non-null object
retweeted_status_id           79 non-null float64
retweeted_status_user_id      79 non-null float64
retweeted_status_timestamp    79 non-null object
expanded_urls                 2073 non-null object
rating_numerator              2073 non-null int64
rating_denominator            2073 non-null int64
name                          2073 non-null object
doggo                         2073 non-null object
floofer                       2073 non-null object
pupper                        2073 non-null object
puppo                         2073 non-null object
jpg_url                       2073 

#### Retweets: Some entries are retweets. We only want to keep the original.

In [24]:
# Identify retweets
df_clean_twitter_archive[df_clean_twitter_archive['retweeted_status_id'].notnull() == True ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
32    885311592912609280                    NaN                  NaN   
67    877611172832227328                    NaN                  NaN   
81    873697596434513921                    NaN                  NaN   
109   867072653475098625                    NaN                  NaN   
129   861769973181624320                    NaN                  NaN   
133   860924035999428608                    NaN                  NaN   
170   851953902622658560                    NaN                  NaN   
171   851861385021730816                    NaN                  NaN   
186   847971574464610304                    NaN                  NaN   
209   842892208864923648                    NaN                  NaN   
215   841833993020538882                    NaN                  NaN   
226   839290600511926273                    NaN                  NaN   
229   838916489579200512                    NaN                  NaN   
238   837012587749474308                    NaN                  NaN   
263   832769181346996225                    NaN                  NaN   
269   832215726631055365                    NaN                  NaN   
270   832040443403784192                    NaN                  NaN   
284   829878982036299777                    NaN                  NaN   
318   824796380199809024                    NaN                  NaN   
327   823269594223824897                    NaN                  NaN   
331   822647212903690241                    NaN                  NaN   
337   821813639212650496                    NaN                  NaN   
346   820446719150292993                    NaN                  NaN   
355   819015337530290176                    NaN                  NaN   
356   819015331746349057                    NaN                  NaN   
361   818588835076603904                    NaN                  NaN   
370   817181837579653120                    NaN                  NaN   
374   816829038950027264                    NaN                  NaN   
380   816014286006976512                    NaN                  NaN   
390   813944609378369540                    NaN                  NaN   
...                  ...                    ...                  ...   
502   794983741416415232                    NaN                  NaN   
504   794355576146903043                    NaN                  NaN   
509   793614319594401792                    NaN                  NaN   
532   791026214425268224                    NaN                  NaN   
535   790723298204217344                    NaN                  NaN   
551   788070120937619456                    NaN                  NaN   
563   786036967502913536                    NaN                  NaN   
578   783347506784731136                    NaN                  NaN   
586   782021823840026624                    NaN                  NaN   
597   780496263422808064                    NaN                  NaN   
598   780476555013349377                    NaN                  NaN   
610   778396591732486144                    NaN                  NaN   
617   777641927919427584                    NaN                  NaN   
620   776819012571455488                    NaN                  NaN   
627   775898661951791106                    NaN                  NaN   
646   772615324260794368                    NaN                  NaN   
656   771171053431250945                    NaN                  NaN   
665   770093767776997377                    NaN                  NaN   
685   766078092750233600                    NaN                  NaN   
704   761750502866649088                    NaN                  NaN   
708   761371037149827077                    NaN                  NaN   
723   759566828574212096                    NaN                  NaN   
727   759159934323924993                    NaN                  NaN   
739   757729163776290825                    NaN         

In [25]:
# Drop retweets
df_clean_twitter_archive.drop(df_clean_twitter_archive[df_clean_twitter_archive['retweeted_status_id'].notnull() == True ].index, inplace=True)

In [26]:
# Check result
df_clean_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2072
Data columns (total 30 columns):
tweet_id                      1994 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     1994 non-null object
source                        1994 non-null object
text                          1994 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1994 non-null object
rating_numerator              1994 non-null int64
rating_denominator            1994 non-null int64
name                          1994 non-null object
doggo                         1994 non-null object
floofer                       1994 non-null object
pupper                        1994 non-null object
puppo                         1994 non-null object
jpg_url                       1994 non

 
 #### Wrong data type: **timestamp** and **retweeted_status_timestamp** should be datetime instead of object.

In [27]:
# Change datetype to datetime64
df_clean_twitter_archive.timestamp = pd.to_datetime(df_clean_twitter_archive.timestamp)
df_clean_twitter_archive.retweeted_status_timestamp = pd.to_datetime(df_clean_twitter_archive.retweeted_status_timestamp)

#### Wrong data type: **in_reply_to_status_id** ,**in_reply_to_user_id**, **retweeted_status_id**, and **retweeted_status_user_id** should be string instead of float64.

In [28]:
# Change datatype to int64
df_clean_twitter_archive['tweet_id'] = df_clean_twitter_archive['tweet_id'].values.astype(np.str)
df_clean_twitter_archive['in_reply_to_status_id'] = df_clean_twitter_archive['in_reply_to_status_id'].values.astype(np.str)
df_clean_twitter_archive['in_reply_to_user_id'] = df_clean_twitter_archive['in_reply_to_user_id'].values.astype(np.str)
df_clean_twitter_archive['retweeted_status_id'] = df_clean_twitter_archive['retweeted_status_id'].values.astype(np.str)
df_clean_twitter_archive['retweeted_status_user_id'] = df_clean_twitter_archive['retweeted_status_user_id'].values.astype(np.str)

# Check results
df_clean_twitter_archive.dtypes

tweet_id                              object
in_reply_to_status_id                 object
in_reply_to_user_id                   object
timestamp                     datetime64[ns]
source                                object
text                                  object
retweeted_status_id                   object
retweeted_status_user_id              object
retweeted_status_timestamp    datetime64[ns]
expanded_urls                         object
rating_numerator                       int64
rating_denominator                     int64
name                                  object
doggo                                 object
floofer                               object
pupper                                object
puppo                                 object
jpg_url                               object
img_num                                int64
p1                                    object
p1_conf                              float64
p1_dog                                  bool
p2        

#### Inconsistent capitalization: Some of the first letter in p1/p2/p3 are capital.

In [29]:
# Lower them all
df_clean_twitter_archive['p1'] = df_clean_twitter_archive['p1'].str.lower()
df_clean_twitter_archive['p2'] = df_clean_twitter_archive['p2'].str.lower()
df_clean_twitter_archive['p3'] = df_clean_twitter_archive['p3'].str.lower()

# Check results
print(df_clean_twitter_archive['p1'].str.isupper().value_counts(), df_clean_twitter_archive['p2'].str.isupper().value_counts(), df_clean_twitter_archive['p3'].str.isupper().value_counts())

False    1994
Name: p1, dtype: int64 False    1994
Name: p2, dtype: int64 False    1994
Name: p3, dtype: int64


#### **twitter_archive**: one variable in four columns (doggo, floofer, pupper, puppo).

In [30]:
# Melt 'doggo', 'floofer', 'pupper', 'puppo' four columns to one column 'dog_stage'
melt_columns = ['doggo', 'floofer', 'pupper', 'puppo']
keep_columns = [c for c in df_clean_twitter_archive.columns.tolist() if c not in melt_columns]
df_clean_twitter_archive = pd.melt(df_clean_twitter_archive, id_vars=keep_columns, value_vars=melt_columns, var_name='stage', value_name='dog_stage')
df_clean_twitter_archive = df_clean_twitter_archive.drop('stage', 1)

# Check results
df_clean_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7976 entries, 0 to 7975
Data columns (total 27 columns):
tweet_id                      7976 non-null object
in_reply_to_status_id         7976 non-null object
in_reply_to_user_id           7976 non-null object
timestamp                     7976 non-null datetime64[ns]
source                        7976 non-null object
text                          7976 non-null object
retweeted_status_id           7976 non-null object
retweeted_status_user_id      7976 non-null object
retweeted_status_timestamp    0 non-null datetime64[ns]
expanded_urls                 7976 non-null object
rating_numerator              7976 non-null int64
rating_denominator            7976 non-null int64
name                          7976 non-null object
jpg_url                       7976 non-null object
img_num                       7976 non-null int64
p1                            7976 non-null object
p1_conf                       7976 non-null float64
p1_dog         

In [31]:
# Drop duplicated rows
df_clean_twitter_archive = df_clean_twitter_archive.drop_duplicates()

In [32]:
# Check the values of 'dog_stage'
df_clean_twitter_archive.dog_stage.value_counts()

None       1994
pupper      212
doggo        74
puppo        23
floofer       8
Name: dog_stage, dtype: int64

#### Name: In name column, many entries do not look like names.

In [33]:
# Replace strange values with None

df_clean_twitter_archive['name'].replace("an",'None', inplace=True)
df_clean_twitter_archive['name'].replace("a",'None', inplace=True)
df_clean_twitter_archive['name'].replace("by",'None', inplace=True)
df_clean_twitter_archive['name'].replace("actually",'None', inplace=True)
df_clean_twitter_archive['name'].replace("old",'None', inplace=True) 
df_clean_twitter_archive['name'].replace("all",'None', inplace=True) 
df_clean_twitter_archive['name'].replace("this",'None', inplace=True) 
df_clean_twitter_archive['name'].replace("very",'None', inplace=True) 
df_clean_twitter_archive['name'].replace("not",'None', inplace=True)
df_clean_twitter_archive['name'].replace("one",'None', inplace=True)
df_clean_twitter_archive['name'].replace("my",'None', inplace=True)
df_clean_twitter_archive['name'].replace("O",'None', inplace=True)
df_clean_twitter_archive['name'].replace("quite","None", inplace=True)
df_clean_twitter_archive['name'].replace("such","None", inplace=True)
df_clean_twitter_archive['name'].replace('the', 'None', inplace=True)
df_clean_twitter_archive['name'].replace("just",'None', inplace=True)
df_clean_twitter_archive['name'].replace("getting",'None', inplace=True) 
df_clean_twitter_archive['name'].replace("infuriating",'None', inplace=True) 

In [34]:
df_clean_twitter_archive['name'].value_counts()

None        774
Cooper       13
Oliver       12
Charlie      12
Lucy         11
Winston       9
Penny         9
Tucker        9
Bo            8
Sadie         8
Lola          7
Milo          7
Toby          7
Daisy         7
Bella         7
Jax           7
Stanley       6
Gus           6
Koda          6
Maggie        6
Scout         6
Finn          5
Derek         5
Chester       5
Clark         5
Winnie        5
Leo           5
Dave          5
Bailey        5
Oscar         5
           ... 
Beckham       1
Jennifur      1
Biden         1
Ralph         1
Puff          1
Oshie         1
Swagger       1
Olaf          1
Juckson       1
Robin         1
Gin           1
Comet         1
Amy           1
Miley         1
Carbon        1
Lassie        1
Howie         1
General       1
Bode          1
Danny         1
Tango         1
Monty         1
Pipsy         1
Tater         1
Jay           1
Spencer       1
Lacy          1
Winifred      1
Cilantro      1
Carter        1
Name: name, Length: 919,

#### Rating: The numerator and denominator columns have strange values.

In [35]:
df_clean_twitter_archive['rating_denominator'].value_counts()

10     2293
50        3
11        2
80        2
7         1
170       1
150       1
130       1
120       1
110       1
90        1
70        1
40        1
20        1
2         1
Name: rating_denominator, dtype: int64

In [36]:
# Select only denominator = 10
df_clean_twitter_archive = df_clean_twitter_archive.query('rating_denominator == 10')
df_clean_twitter_archive['rating_denominator'].value_counts()

10    2293
Name: rating_denominator, dtype: int64

In [37]:
df_clean_twitter_archive['rating_numerator'].value_counts()

12      527
10      478
11      469
13      308
9       167
8       110
7        59
14       50
5        36
6        33
3        20
4        15
2         9
1         4
0         2
27        2
75        1
26        1
1776      1
420       1
Name: rating_numerator, dtype: int64

In [38]:
# Selet only numerator <= 20 and > 0 
df_clean_twitter_archive = df_clean_twitter_archive.query('rating_numerator <= 20 and rating_numerator > 0')
df_clean_twitter_archive['rating_numerator'].value_counts()

12    527
10    478
11    469
13    308
9     167
8     110
7      59
14     50
5      36
6      33
3      20
4      15
2       9
1       4
Name: rating_numerator, dtype: int64

In [39]:
df_clean_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2285 entries, 0 to 6795
Data columns (total 27 columns):
tweet_id                      2285 non-null object
in_reply_to_status_id         2285 non-null object
in_reply_to_user_id           2285 non-null object
timestamp                     2285 non-null datetime64[ns]
source                        2285 non-null object
text                          2285 non-null object
retweeted_status_id           2285 non-null object
retweeted_status_user_id      2285 non-null object
retweeted_status_timestamp    0 non-null datetime64[ns]
expanded_urls                 2285 non-null object
rating_numerator              2285 non-null int64
rating_denominator            2285 non-null int64
name                          2285 non-null object
jpg_url                       2285 non-null object
img_num                       2285 non-null int64
p1                            2285 non-null object
p1_conf                       2285 non-null float64
p1_dog         

#### Remove columns which are not required for the analysis and store the dataframe to twitter_archive_master.csv

In [40]:
df_clean_twitter_archive.to_csv('twitter_archive_master.csv')

## Analyze data

In [41]:
df = pd.read_csv('twitter_archive_master.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2285 entries, 0 to 2284
Data columns (total 28 columns):
Unnamed: 0                    2285 non-null int64
tweet_id                      2285 non-null int64
in_reply_to_status_id         25 non-null float64
in_reply_to_user_id           25 non-null float64
timestamp                     2285 non-null object
source                        2285 non-null object
text                          2285 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null float64
expanded_urls                 2285 non-null object
rating_numerator              2285 non-null int64
rating_denominator            2285 non-null int64
name                          2285 non-null object
jpg_url                       2285 non-null object
img_num                       2285 non-null int64
p1                            2285 non-null object
p1_conf                       2285 non-